submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,197 @@
+## Globalnet 2.0: Enhancements to Globalnet Implementation
+
+## Summary
+
+Currently, Submariner Globalnet implementation in [release 0.8](https://github.com/submariner-io/releases/releases/tag/v0.8.0)
+annotates every Pod, Service and Node in the cluster with a globalIp to support Pod to remote
+Service connectivity in the following manner.
+
+Once a Pod is scheduled and enters Running state, it is annotated with a globalIp. To support
+the connectivity, Globalnet programs the necessary egress rules on the active Gateway Node of
+the cluster where the source IP of the Pod is modified to the corresponding globalIp. This
+translation enables us to uniquely identify the traffic from the Pod, when it reaches the
+destination cluster.
+
+Similarly, a ClusterIP Service is annotated with a globalIp so that remote Pods can communicate
+with it using its globalIp. Globalnet programs the necessary ingress rules on the active Gateway
+node of the cluster, where the service resides. The ingress rules enable any incoming traffic
+destined to the globalIp of the Service to be redirected to the corresponding ClusterIP Service.
+
+In order to support connectivity from the Host Network to the remote services, Globalnet also
+annotates every node in the cluster with a unique globalIp.
+
+While this works fine for small clusters, where the number of Pods and Services are limited, for
+large clusters it poses scalability and performance issues as we would eventually run out of
+globalIps because of annotating every Pod/Service/Node object. Also, in this design, Pod to Pod
+connectivity as well as Headless Services are not supported.
+
+## Proposal
+
+This enhancement proposes the following changes to make Globalnet implementation scale better.
+
+1. By default, every cluster will be assigned a single globalIp which will be used as egress IP
+   for any cross-cluster communication.
+2. Provision to allocate a globalIp at the namespace/project level which takes precedence over
+   the globalIp allocated to the cluster.
+3. Provision to allocate a globalIp to a Pod (or a set of Pods) which has higher precedence
+   over globalIPs at namespace level. If there are multiple policies, users can specify a priority
+   field for each of the policy.
+4. Instead of annotating every ClusterIP Service in the cluster, only exported Services will be
+   annotated with a globalIp.
+5. For Headless Services, necessary ingress rules will be programmed to support connectivity to
```

Comment:
IMHO egress IP should be kept independent of ingress IP.

For example, for the external network use cases, where people want a consistent IP for pods (same IP for a namespace or set of pods, or cluster...) ... that will make varying IPS for the pods as soon as they restart.

So for that I propose that this would only talk of the ingress IP.

Reply:
Yes, but have you forgotten all the meetings about external networks?

We are trying to satisfy the 2nd requirement that @mkimuram found, by separating those details in the design.

This is important for them because they use source IP for authentication (typical of mysql  for example.)

Reply:
And I think it's natural, I don't believe the egress IP of a POD needs to be modified by a service definition,

Reply:
What if... 

1) When there is no specific GlobalnetEgressIP, the egress IP will be the one of the pod in the headless services
2) If there is one GlobalnetEgressIP requested, that will take precedence?

That way we can satisfy both cases?

It would take precedence over the cluster global IP :)

I make harder design, but I'm trying to satisfy all possible presented use cases.

Reply:
AFAIU the 2nd requirement from @mkimuram is

When a Pod (aka Client) tries to talk to some External application, the sourceIP seen by External application should not change even when the Pod restarts. 
This will allow External application to use SG/ACLs to allow traffic from a particular IP.

To achive this use-case, in the new design we have the following possibilities

1. User can explicitly request for a GlobalnetEgressIP at the namespace level. But the drawback is that all the Pods in that namespace will have access to the External application.

2. User can explicitly request for a GlobalnetEgressIP for the particular Pod using PodSelectors. Assuming that there is only a single client Pod, user can even specify the globalIP that he wants the Pod to use.
   This will enable Globalnet to assign the same globalIP to the Pod even when the Pod restarts.
   
3. User creates headless service with backend Pods and always uses DNS/LH to resolve the latest IPs and configure appropriate ACL/SG rules on the external application.

Reply:
In my awful diagram that we shared over bluejeans:

```

GlobalnetEgressIP:
   PodSelector:
      app=http1
                
   Status:
      GlobalIP: 169.254.0.11
               
               
  http-1234  10.0.0.1     <------  169.254.0.12
  http-3333  10.0.0.2     <------  169.254.0.13    -- 169.254.0.11 ----->      ExternalNetwork ------->       10.1.0.3   (database-A)   ok, so PWD OK, IP==169.254.0.11 --> login ok
  http-4444  10.0.0.3     <------  169.254.0.14                                                                                         ok, so PWD OK, IP!=169.254.0.11 --> login FAILED
  
  
          Service:
            selector:
               app=http1
            ClusterIP: None
            
            
   If there is no GlobalnetEgressIP selector on those pods
  
                 
  http-1234  10.0.0.1     <------  169.254.0.12    -- 169.254.0.12 ----->
  http-3333  10.0.0.2     <------  169.254.0.13    -- 169.254.0.13 ----->    ExternalNetwork/Other cluster ------->       10.1.0.3  
  http-4444  10.0.0.3     <------  169.254.0.14    -- 169.254.0.14 ----->                                                                                                      
  
  
  
 
```

Reply:
Done